[jfrog/distribution] Distribution Chart Results in CrashLoopBackOff

**Is this a request for help?**: No

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Version of Helm and Kubernetes**:

Helm Version:

```
version.BuildInfo{Version:"v3.1.2", GitCommit:"d878d4d45863e42fd5cff6743294a11d28a9abce", GitTreeState:"clean", GoVersion:"go1.13.8"}
```

Kubernetes Version:

```
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.9", GitCommit:"2e808b7cb054ee242b68e62455323aa783991f03", GitTreeState:"clean", BuildDate:"2020-01-18T23:33:14Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.11-gke.5", GitCommit:"a5bf731ea129336a3cf32c3375317b3a626919d7", GitTreeState:"clean", BuildDate:"2020-03-31T02:49:49Z", GoVersion:"go1.12.17b4", Compiler:"gc", Platform:"linux/amd64"}
```


**Which chart**: jfrog/distribution


**What happened**: The statefulset pod of jfrog/distribution results in `CrashLoopBackOff` of the `distribution-0` pod. Specifically, the ReadinessProbe and LivenessProbe fail which causes the container `distribution` to be restarted. I do not see any errors when running `kubectl logs distribution-0 -c distribution`. However, I do see errors when running `kubectl logs distribution-0 -c distributor` as follows:

```
2020-04-20T14:28:24.471Z [jfdr ] [INFO ] [1ded62c53a4b955 ] [c.j.b.d.m.HeartbeatJob:43     ] [main                ] - Scheduled Heartbeat job with the following period: 5 SECONDS
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.vmplugin.v7.Java7$1 (file:/opt/jfrog/distribution/app/distributor/lib/distributor.jar) to constructor java.lang.invoke.MethodHandles$Lookup(java.lang.Class,int)
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.vmplugin.v7.Java7$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[LOG BOOTSTRAP] Using the value of JF_PRODUCT_HOME environment variable for distribution home path
[LOG BOOTSTRAP] Using the value of JF_PRODUCT_HOME environment variable for distribution home path
2020-04-20T14:28:40.124Z [jfdr ] [INFO ] [1ded62c53a4b955 ] [DistributionApiTokenService:58] [main                ] - Searching for Distribution api token at /opt/jfrog/distribution/var/etc/distributor/security/token
2020-04-20T14:28:40.125Z [jfdr ] [INFO ] [1ded62c53a4b955 ] [DistributionApiTokenService:61] [main                ] - Distribution api token file not found.
2020-04-20T14:28:40.125Z [jfdr ] [ERROR] [1ded62c53a4b955 ] [DistributionApiTokenService:48] [main                ] - Distribution api token is blank.
2020-04-20T14:28:40.125Z [jfdr ] [WARN ] [1ded62c53a4b955 ] [c.j.b.d.DistributorRunner:44  ] [main                ] - Distribution must be up and bound to an authentication provider before additional tasks can be performed.
```


**What you expected to happen**: I expect it to work


**How to reproduce it** (as minimally and precisely as possible):

Run the following command:

```
helm upgrade --install distribution -f distribution.yaml \
--set distribution.joinKey="${JOIN_KEY}" \
--set distribution.jfrogUrl="${JFROG_URL}" \
--set distribution.masterKey=${DISTRIBUTION_MASTER_KEY} \
jfrog/distribution
```

My `distribution.yaml` is as follows:

```
replicaCount: 4

redis:
  resources:
    requests:
      memory: "512Mi"
      cpu: "1"
    limits:
      memory: "2Gi"
      cpu: "2"

distribution:
  resources:
    requests:
      memory: "2Gi"
      cpu: "1"
    limits:
      memory: "6Gi"
      cpu: "4"
  javaOpts:
    xms: "2g"
    xmx: "6g"

distributor:
  resources:
    requests:
      memory: "2Gi"
      cpu: "1"
    limits:
      memory: "6Gi"
      cpu: "4"
  javaOpts:
    xms: "2g"
    xmx: "6g"
```


**Anything else we need to know**:

--- a/stable/distribution/CHANGELOG.md
+++ b/stable/distribution/CHANGELOG.md
@@ -1,6 +1,10 @@
 # JFrog Distribution Chart Changelog
 All changes to this project chart be documented in this file.
 
+## [5.1.3] - April 24, 2020
+* Add support to db existing secrets
+* Fix broken support if `postgresql.postgresqlPassword` not explicitly set
+
 ## [5.1.2] - April 20, 2020
 * Bump router version to `1.3.0`
 * Fix broken support if `redis.password` not explicitly set

--- a/stable/distribution/Chart.yaml
+++ b/stable/distribution/Chart.yaml
@@ -2,7 +2,7 @@ apiVersion: v1
 name: distribution
 home: https://jfrog.com/platform/
 description: A Helm chart for JFrog Distribution
-version: 5.1.2
+version: 5.1.3
 appVersion: 2.2.0
 sources:
 - https://bintray.com/jfrog/product/distribution/view

--- a/stable/distribution/README.md
+++ b/stable/distribution/README.md
@@ -18,7 +18,7 @@ This chart does the following:
 - Default StorageClass set to allow services using the default StorageClass for persistent storage
 - A running Artifactory Enterprise Plus
 - [Kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/) installed and setup to use the cluster
-- [Helm](https://helm.sh/) installed and setup to use the cluster (helm init)
+- [Helm](https://helm.sh/) v2 or v3 installed
 
 ## Installing the Chart
 
@@ -298,12 +298,15 @@ The following table lists the configurable parameters of the distribution chart 
 | `database.type`                                 | External database type (`postgresql`)                                  | `postgresql`                                      |
 | `database.driver`                               | External database driver                                               | 
 `org.postgresql.Driver`                           |
-| `database.url`                                  | External database url                                                  |
-` `                                               |
-| `database.user`                                 | External database user                                                 |
-` `                                               |
-| `database.password`                             | External database password                                             |
-` `                                               |
+| `database.url`                                  | External database connection URL                  |                                         |
+| `database.user`                                 | External database username                         |                                         |
+| `database.password`                             | External database password                         |                                         |
+| `database.secrets.user.name`                    | External database username `Secret` name           |                                         |
+| `database.secrets.user.key`                     | External database username `Secret` key            |                                         |
+| `database.secrets.password.name`                | External database password `Secret` name           |                                         |
+| `database.secrets.password.key`                 | External database password `Secret` key            |                                         |
+| `database.secrets.url.name`                     | External database url `Secret` name                |                                         |
+| `database.secrets.url.key`                      | External database url `Secret` key                 |                                         |
 | `distribution.joinKey`                          | Distribution Join Key . Mandatory                                      | ` `                                                                |
 | `distribution.serviceId`                        | Distribution service ID                                                | ` `                                                                |
 | `distribution.env.artifactoryUrl`               | Distribution Environment Artifactory URL                               | ` `                                                                |

--- a/stable/distribution/templates/database-secrets.yaml
+++ b/stable/distribution/templates/database-secrets.yaml
@@ -1,8 +1,8 @@
-{{- if .Values.database.password }}
+{{- if and  (not .Values.database.secrets) (not .Values.postgresql.enabled) }}
 apiVersion: v1
 kind: Secret
 metadata:
-  name: {{ template "distribution.fullname" . }}-database-connection
+  name: {{ template "distribution.fullname" . }}-database-creds
   labels:
     app: {{ template "distribution.name" . }}
     chart: {{ template "distribution.chart" . }}
@@ -11,5 +11,13 @@ metadata:
     component: {{ .Values.distribution.name }}
 type: Opaque
 data:
+  {{- if .Values.database.url }}
+  db-url: {{ .Values.database.url | b64enc | quote }}
+  {{- end }}
+  {{- if .Values.database.user }}
+  db-user: {{ .Values.database.user | b64enc | quote }}
+  {{- end }}
+  {{- if .Values.database.password }}
   db-password: {{ .Values.database.password | b64enc | quote }}
+  {{- end }}
 {{- end }}

--- a/stable/distribution/templates/distribution-statefulset.yaml
+++ b/stable/distribution/templates/distribution-statefulset.yaml
@@ -146,6 +146,45 @@ spec:
           {{- end }}
             /opt/jfrog/distribution/app/bin/wrapper.sh;
         env:
+        {{- if or .Values.database.secrets.user .Values.database.user }}
+        - name: JF_SHARED_DATABASE_USERNAME
+          valueFrom:
+            secretKeyRef:
+        {{- if .Values.database.secrets.user }}
+              name: {{ tpl .Values.database.secrets.user.name .}}
+              key: {{ tpl .Values.database.secrets.user.key .}}
+        {{- else if .Values.database.user }}
+              name: {{ template "distribution.fullname" . }}-database-creds
+              key: db-user
+        {{- end }}
+      {{- end }}
+      {{ if or .Values.database.secrets.password .Values.database.password .Values.postgresql.enabled }}
+        - name: JF_SHARED_DATABASE_PASSWORD
+          valueFrom:
+            secretKeyRef:
+        {{- if .Values.database.secrets.password }}
+              name: {{ tpl .Values.database.secrets.password.name . }}
+              key: {{ tpl .Values.database.secrets.password.key . }}
+        {{- else if .Values.database.password }}
+              name: {{ template "distribution.fullname" . }}-database-creds
+              key: db-password
+        {{- else if .Values.postgresql.enabled }}
+              name: {{ .Release.Name }}-postgresql
+              key: postgresql-password
+        {{- end }}
+      {{- end }}
+      {{- if or .Values.database.secrets.url .Values.database.url }}
+        - name: JF_SHARED_DATABASE_URL
+          valueFrom:
+            secretKeyRef:
+        {{- if .Values.database.secrets.url }}
+              name: {{ tpl .Values.database.secrets.url.name .}}
+              key: {{ tpl .Values.database.secrets.url.key .}}
+        {{- else if .Values.database.url }}
+              name: {{ template "distribution.fullname" . }}-database-creds
+              key: db-url
+        {{- end }}
+      {{- end }}
         - name: JF_SHARED_REDIS_PASSWORD
           valueFrom:
             secretKeyRef:
@@ -193,6 +232,45 @@ spec:
           {{- end }}
             /opt/jfrog/distribution/app/bin/wrapper.sh;
         env:
+        {{- if or .Values.database.secrets.user .Values.database.user }}
+        - name: JF_SHARED_DATABASE_USERNAME
+          valueFrom:
+            secretKeyRef:
+        {{- if .Values.database.secrets.user }}
+              name: {{ tpl .Values.database.secrets.user.name .}}
+              key: {{ tpl .Values.database.secrets.user.key .}}
+        {{- else if .Values.database.user }}
+              name: {{ template "distribution.fullname" . }}-database-creds
+              key: db-user
+        {{- end }}
+      {{- end }}
+      {{ if or .Values.database.secrets.password .Values.database.password .Values.postgresql.enabled }}
+        - name: JF_SHARED_DATABASE_PASSWORD
+          valueFrom:
+            secretKeyRef:
+        {{- if .Values.database.secrets.password }}
+              name: {{ tpl .Values.database.secrets.password.name . }}
+              key: {{ tpl .Values.database.secrets.password.key . }}
+        {{- else if .Values.database.password }}
+              name: {{ template "distribution.fullname" . }}-database-creds
+              key: db-password
+        {{- else if .Values.postgresql.enabled }}
+              name: {{ .Release.Name }}-postgresql
+              key: postgresql-password
+        {{- end }}
+      {{- end }}
+      {{- if or .Values.database.secrets.url .Values.database.url }}
+        - name: JF_SHARED_DATABASE_URL
+          valueFrom:
+            secretKeyRef:
+        {{- if .Values.database.secrets.url }}
+              name: {{ tpl .Values.database.secrets.url.name .}}
+              key: {{ tpl .Values.database.secrets.url.key .}}
+        {{- else if .Values.database.url }}
+              name: {{ template "distribution.fullname" . }}-database-creds
+              key: db-url
+        {{- end }}
+      {{- end }}
         - name: JF_SHARED_REDIS_PASSWORD
           valueFrom:
             secretKeyRef:

--- a/stable/distribution/values.yaml
+++ b/stable/distribution/values.yaml
@@ -72,7 +72,7 @@ serviceAccount:
 
 # PostgreSQL
 ## Configuration values for the PostgreSQL dependency sub-chart
-## ref: https://github.com/kubernetes/charts/blob/master/stable/postgresql/README.md
+## ref: https://github.com/bitnami/charts/blob/master/bitnami/postgresql/README.md
 postgresql:
   enabled: true
   image:

--- a/stable/distribution/values.yaml
+++ b/stable/distribution/values.yaml
@@ -102,24 +102,28 @@ postgresql:
   affinity: {}
 
 ## If NOT using the PostgreSQL in this chart (postgresql.enabled=false),
-## specify custom database details here or leave empty and Artifactory will use embedded derby
+## specify custom/external database details here
 database:
   type: postgresql
   driver: org.postgresql.Driver
-  url:
   ## If you would like this chart to create the secret containing the db
   ## password, use these values
+  url:
   user:
   password:
   ## If you have existing Kubernetes secrets containing db credentials, use
   ## these values
   secrets: {}
   #  user:
-  #    name: "database-creds"
+  #    name: "dis-database-creds"
   #    key: "db-user"
   #  password:
-  #    name: "database-creds"
+  #    name: "dis-database-creds"
   #    key: "db-password"
+  #  url:
+  #    name: "dis-database-creds"
+  #    key: "db-url"
+
 
 ## Configuration values for the redis dependency
 ## ref: https://github.com/helm/charts/blob/master/stable/redis/README.md
@@ -227,14 +231,10 @@ distribution:
         type: "postgresql"
         driver: "org.postgresql.Driver"
         username: "{{ .Values.postgresql.postgresqlUsername }}"
-        password: "{{ .Values.postgresql.postgresqlPassword }}"
         url: "postgresql://{{ .Release.Name }}-postgresql:{{ .Values.postgresql.service.port }}/{{ .Values.postgresql.postgresqlDatabase }}"
       {{ else }}
         type: "{{ .Values.database.type }}"
         driver: "{{ .Values.database.driver }}"
-        url: "{{ .Values.database.url }}"
-        username: "{{ .Values.database.user }}"
-        password: "{{ .Values.database.password }}"
       {{- end }}
 
   service:

